### PR TITLE
rgw_sal_motr: fix data leak on put error

### PR DESCRIPTION
**Pull request by andriytk**
Currently, if there is an error on object metadata dix put
operation at the end of object PUT operation - the written
motr object data is not cleaned up and is leaked.

Solution: cleanup motr object on metadata put error.

Also, don't delete the old object if the new object put has failed.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_sal_motr.cc [hunk 1]
```diff
@@ -2484,10 +2484,12 @@ int MotrAtomicWriter::complete(size_t accounted_size, const std::string& etag,
   if (rc == 0)
     store->get_obj_meta_cache()->put(dpp, obj.get_key().to_str(), bl);
 
-  if (old_obj.get_bucket()->get_info().versioning_status() != BUCKET_VERSIONED) {
-    // Delete old object data if exists.
-    old_obj.delete_mobj(dpp);
-  }
+  if (rc == 0 && is_versioned)
```
Comment (sjain09) [inline src/rgw/rgw_sal_motr.cc:2487]: `is_versioned` does not necessarily say whether bucket versioning is enabled or not. I think the previous check for bucket versioning was more appropriate. We can have additional check in the previous if statement itself.

Comment (andriytk) [reply src/rgw/rgw_sal_motr.cc:2487]: Ok, two questions then:
1) If we need to check bucket - why not to use `bucket->versioning_enabled()`?
2) Should we set `is_versioned` flag above using the same method?

Comment (andriytk) [reply src/rgw/rgw_sal_motr.cc]: See the updated version of the patch.

**Review by sjain09 — APPROVED** [review]
Looks good.

**Pull request merged**